skorch/probabilistic.py

```diff
@@ -0,0 +1,941 @@
+        input data.
+
+        The outputs from ``self.module_.forward`` are gathered on the
+        compute device specified by ``device`` and then concatenated
+        using PyTorch :func:`~torch.cat`. If multiple outputs are
+        returned by ``self.module_.forward``, each one of them must be
+        able to be concatenated this way.
+
+        Notes
+        -----
+        For Gaussian Process modules, the return value of the module is a
+        distribution. These distributions are collected in a list (which may
+        only contain a single element if just one batch was used). Distributions
+        *cannot* be concatenated. Therefore, this method will just return the
+        list of distributions.
+
+        Parameters
+        ----------
+        X : input data, compatible with skorch.dataset.Dataset
+          By default, you should be able to pass:
+
+            * numpy arrays
+            * torch tensors
+            * pandas DataFrame or Series
+            * scipy sparse CSR matrices
+            * a dictionary of the former three
+            * a list/tuple of the former three
+            * a Dataset
+
+          If this doesn't work with your data, you have to pass a
+          ``Dataset`` that can deal with the data.
+
+        training : bool (default=False)
+          Whether to set the module to train mode or not.
+
+        device : string (default='cpu')
+          The device to store each inference result on.
+          This defaults to CPU memory since there is genereally
+          more memory available there. For performance reasons
+          this might be changed to a specific CUDA device,
+          e.g. 'cuda:0'.
+
+        Returns
+        -------
+        y_infer : list of gpytorch.distributions.Distribution
+          A list of distributions as generated by the module. The number of
+          elements in this list will depend on the sample size of X and the
+          batch size of the estimator.
+
+        """
+        y_infer = list(self.forward_iter(X, training=training, device=device))
+        return y_infer
+
+    def predict_proba(self, X):
+        raise AttributeError("'predict_proba' is not implemented for {}".format(
+            self.__class__.__name__
+        ))
+
+    def sample(self, X, n_samples, axis=-1):
+        """Return samples conditioned on input data.
+
+        The GP doesn't need to be fitted but it must be initialized.
+
+        X : input data
+          The samples where the GP is evaluated.
+
+        n_samples : int
+          The number of samples to return
+
+        axis : int (default=-1)
+          The concatenation axis of the samples. Since samples can come in
+          batches, they must be concatenated.
+
+        Returns
+        -------
+        samples : torch.Tensor
+          Samples from the posterior distribution.
+
+        """
+        self.check_is_fitted()
+        samples = [p.sample(torch.Size([n_samples])) for p in self.forward_iter(X)]
```

Comment:
Obviously not 100% familiar with how skorch's internals work, but assuming `p` is a `MultivariateNormal`, is there any reason to call `sample` instead of `rsample` here? Outright enforcing that sampling must be done in a `torch.no_grad()` context seems unnecessary at first glance, and a common use case for GPs looks something like: `x -> GP -> p(y|D) -> samples of y -> average some function of y over the samples -> differentiate the average with respect to x`

Reply:
The main reason I used `sample` was because `rsample` was not supported for all distributions, e.g. `Bernoulli`. I didn't think of the possibility of using the sample gradients. I'll make a switch to `rsample` if the used distribution supports it and fall back on `sample` otherwise.